srk/src/weakSolver.ml

```diff
@@ -0,0 +1,196 @@
+
+include Log.Make(struct let name = "srk.weaksolver" end)
+
+
+let is_sat_model srk phi =
+  (* Use normalize in quantifier.ml to compute the prenex normal form and get the existentials *)
+  let phi = eliminate_ite srk phi in
+  let phi =
+    SrkSimplify.simplify_terms srk phi
+  in
+  let quantifiers, phi = Quantifier.normalize srk phi in
+  logf "formula: %a" (Formula.pp srk) phi;
+  (* This module should be sound wrt the theory, other things like x > 0 <=> x >= 1 happen in other modules *)
+  (* cannot rewrite into negation normal form past equalities, since
+     not (p = 0) is not equiv to p > 0 or p < 0
+     since don't have total order *)
+  assert (BatList.for_all (fun quant -> match quant with `Exists, _ -> true | _ -> false) quantifiers = true);
+
+  let solver = Smt.mk_solver ~theory:"QF_LIRA" srk in
+  let uninterp_phi =
+    rewrite srk
+      ~down:(nnf_rewriter srk)
+      ~up:(Nonlinear.uninterpret_rewriter srk)
+      phi
+  in
+  logf "Uninterpreted formula: %a" (Formula.pp srk) uninterp_phi;
+  (* nonlinear is a map from symbols to term/formula *)
+  let (lin_phi, nonlinear) = SrkSimplify.purify srk uninterp_phi in
+  logf "Linear formula skeleton: %a" (Formula.pp srk) lin_phi;
+  (* Creating a list of relationships between symbols and their meanings *)
+  let nonlinear_defs =
+    Symbol.Map.enum nonlinear
+    /@ (fun (symbol, expr) ->
+        match Expr.refine srk expr with
+        | `Term t -> mk_eq srk (mk_const srk symbol) t
+        | `Formula phi -> mk_iff srk (mk_const srk symbol) phi)
+    |> BatList.of_enum
+  in
+  let nonlinear = Symbol.Map.map (Nonlinear.interpret srk) nonlinear in
+  let rec replace_defs_term term =
+    substitute_const
+      srk
+      (fun x ->
+         try replace_defs_term (Symbol.Map.find x nonlinear)
+         with Not_found -> mk_const srk x)
+      term
+  in
+  let replace_defs =
+    substitute_const
+      srk
+      (fun x ->
+         try replace_defs_term (Symbol.Map.find x nonlinear)
+         with Not_found -> mk_const srk x)
+  in
+  Smt.Solver.add solver [lin_phi];
+  Smt.Solver.add solver nonlinear_defs;
+
+  (* TODO: adding lemmas. Using preduce in Groebner basis to get a subset of atoms whose conjunction is unsat. *)
+   let rec go () =
+    match Smt.Solver.get_model solver with
+    | `Unsat -> `Unsat
+    | `Unknown -> `Unknown
+    | `Sat model ->
+      match Interpretation.select_implicant model lin_phi with
+      | None -> assert false
+      | Some implicant ->
+        (* The formulas should be atoms, should be able to destruct to get t < c, t <= c, t = c *)
+        let atoms = List.map replace_defs implicant in
+        let rec process_atoms l cs geqs eqs ineqs =
+          match l with
+          | [] -> (cs, geqs, eqs, ineqs)
+          | h :: t ->
+              logf "Processing atom: %a" (Formula.pp srk) h;
+              match (Interpretation.destruct_atom srk h) with
+              `Comparison (`Eq, a, b) ->
+              CS.admit_term cs a;
```

Comment:
Coordinate systems aren't needed here -- can add a term-to-polynomial conversion function to polynomial.ml (like the linear variant in linear.ml)